pyflakes/checker.py

```diff
@@ -952,21 +968,27 @@ def IMPORT(self, node):
 
     def IMPORTFROM(self, node):
         if node.module == '__future__':
-            if not self.futuresAllowed:
+            # __future__ can only appear in module/doctest scope and
+            # should not have been disabled already in handleNode and
+            # the scope must not have any other type of binding.
+            if (not isinstance(self.scope, ModuleScope) or
+                    not self.scope._futures_allowed or
+                    any(not isinstance(binding, FutureImportation)
+                        for binding in self.scope.values())):
+                self.scope._futures_allowed = False
```

Comment:
Can this (and the very similar logic above) be broken up to make it easier to digest please?

And maybe it can be consolidated. If I'm following the logic correctly, the only reason we can't simply check `scope._futures_allowed` here is that above, where its set in `handleNode()`, there's a possibility there was some `from foo import bar` or similar. That would have met the `isinstance(node, ast.ImportFrom)` condition, and so would not have set `self.scope._futures_allowed = False`

It would seem though that if we extend that condition just a bit, to match only `from __future__ ...`, then we can have the complete logic for determining if `__future__` imports are allowed in just one place. What do you think?